src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp

```diff
@@ -237,37 +533,181 @@ void CertificateTableImpl::Finish()
+CHIP_ERROR CertificateTableImpl::HasRootCertificateEntry(FabricIndex fabric_index, TLSCAID id)
 {
-    CertificateId id(certificate_id);
+    CertificateId localId(id);
     EntryIndex unused;
     return mRootCertificates.FindTableEntry(fabric_index, id, unused);
 }
 
-CHIP_ERROR CertificateTableImpl::GetClientCertificateEntry(FabricIndex fabric_index, TLSCCDID certificate_id,
-                                                           BufferedClientCert & entry)
+CHIP_ERROR CertificateTableImpl::IterateRootCertificates(FabricIndex fabric, BufferedRootCert & store,
+                                                         IterateRootCertFnType iterateFn)
 {
-    CertificateId id(certificate_id);
-    return mClientCertificates.GetTableEntry(fabric_index, id, entry.mCert, GetBuffer(entry));
+    return mRootCertificates.IterateEntries(fabric, GetBuffer(store), [&](auto & iter) {
+        TableEntryDataConvertingIterator<CertificateId, RootCertStruct> innerIter(iter);
+        return iterateFn(innerIter);
+    });
+}
+
+CHIP_ERROR CertificateTableImpl::RemoveRootCertificate(FabricIndex fabric, TLSCAID id)
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    GlobalCertificateData globalData(mEndpointId);
+    ReturnErrorOnFailure(globalData.Load(mStorage));
+    return globalData.RemoveRootCertificate(*mStorage, mRootCertificates, fabric, id);
 }
 
-CHIP_ERROR CertificateTableImpl::HasClientCertificateEntry(FabricIndex fabric_index, TLSCCDID certificate_id)
+CHIP_ERROR CertificateTableImpl::GetRootCertificateCount(FabricIndex fabric, uint8_t & outCount)
 {
-    CertificateId id(certificate_id);
+    return mRootCertificates.GetFabricEntryCount(fabric, outCount);
+}
+
+CHIP_ERROR CertificateTableImpl::PrepareClientCertificate(FabricIndex fabric, const ByteSpan & nonce, TLSCCDID & id,
+                                                          MutableByteSpan & csr, MutableByteSpan & nonceSignature)
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    // TODO(gmarcosb): Use PSAKeyAllocator instead
+    KeyStruct * nextAvailable = nullptr;
+    for (auto & pending : mPendingClientCerts)
+    {
+        if (!pending.certificateId.HasValue())
+        {
+            nextAvailable = &pending;
+        }
+    }
+
+    if (nextAvailable == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    auto & keyPair = nextAvailable->key;
+
+    // Find a usable ID
+    GlobalCertificateData globalData(mEndpointId);
+    ReturnErrorOnFailure(globalData.Load(mStorage));
+
+    // Update the next ID
+    TLSCCDID localId;
+    ReturnErrorOnFailure(globalData.GetNextClientCertificateId(fabric, localId));
+    ReturnErrorOnFailure(globalData.Save(mStorage));
+
+    ReturnErrorOnFailure(keyPair.Initialize(Crypto::ECPKeyTarget::ECDSA));
+    nextAvailable->certificateId.SetValue(localId);
+    nextAvailable->endpoint = mEndpointId;
+    id                      = localId;
+
+    return CHIP_NO_ERROR;
+}
```

Comment:
I think you need to populate the `csr` [out param] here. The `nonce` [in] and `nonceSignature` [out] seem to be unused as well; only the CCD ID is being allocated in this function.